portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -14,45 +14,82 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
 import com.google.appengine.api.users.UserService;
 import com.google.appengine.api.users.UserServiceFactory;
 import com.google.gson.Gson;
-import com.google.sps.data.User;
+import com.google.sps.data.UserAuthenticationData;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that creates login or logout URL and sends it as response. */
+/**
+ * Servlet that creates login or logout URL and sends it as response.
+ */
 @WebServlet("/login")
 public class LoginServlet extends HttpServlet {
 
+  /**
+   * Instantiates a UserAuthenticationData object and returns it in JSON format.
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
+    
+    // Create variables needed to instantiate a UserAuthenticationData object.
     Boolean loggedIn = false;
     String authenticationUrl = ""; // URL used for either login or logout.
     String userEmail = "";
-    String redirectUrl = "/"; // Both login and logout redirect to the same URL
+    String nickname = "";
+    String redirectUrl = "/"; // Both login and logout redirect to the same URL.
 
+    // Define values for user variables depending on login status.
     if (userService.isUserLoggedIn()) {
-      // Create a new user with logout URL and email.
       loggedIn = true;
-      authenticationUrl = userService.createLogoutURL(redirectUrl);
       userEmail = userService.getCurrentUser().getEmail();
+      nickname = getUserNickname(userService.getCurrentUser().getUserId());
```

Comment:
Don't forget to update this spot at well! You should be able to use userService.getCurrentUser().getNickname() just like in the other servlets.

Reply:
This one is a bit different than the other functions, and for some reason, when using userService.getCurrentUser().getNickname() the nickname gets overwritten with the email, so it stops working as intended.

Reply:
I tried changing logic a bit to see if I could make it work but I get the same results.

Reply:
Ah after reading the API doc (https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/users/User.html#getnickname) my hunch is that the default value of getNickname() is the user's email. So since the nickname is not specifically initialized yet, it just returns the user's email instead. 

There a few options you can consider:
One option is to use userService.getCurrentUser.getNickname and then change your if statement to check if nickname == username. However, this has one odd edge case where someone may want their nickname to be their email (in which case it'll keep telling them to set a nickname - if that happens you could leave a dialog that says "Nickname cannot be the same as the Email)

I'll leave it up to your discretion - I think it would be preferable to utilize the API, but if you'd rather not limit their options for nicknames, you can leave it as-is.

Reply:
> my hunch is that the default value of getNickname() is the user's email.

Ooh, could be the case! I'll implement that first option, and try to add that dialogue if it doesn't take much time.

Reply:
I created the Utils class!